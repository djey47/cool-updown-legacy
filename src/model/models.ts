import { CronJob } from 'cron';
import { TypedResponse } from './express';

export interface ApiRequest {
  params: {
    serverId?: string;
  };
}

export interface ServerState {
  startedAt?: Date;
  stoppedAt?: Date;
  isScheduleEnabled?: boolean;
  onJob?: CronJob;
  offJob?: CronJob;
}

export interface AppState {
  isScheduleEnabled?: boolean;
  onJob?: CronJob;
  offJob?: CronJob;
  startedAt?: Date;
  servers: ServerState[];
}

export interface ServerConfig {
  url?: string;
  network?: {
    broadcastIpAddress: string;
    hostname: string;
    macAddress: string;
  };
  ssh?: {
    keyPath?: string;
    offCommand: string;
    password?: string;
    port?: number;
    user?: string;
  };
  schedule?: {
    enabled: boolean;
    on: Schedule;
    off: Schedule; 
  };
}

export interface AppConfig {
  app?: {
    port?: number;
    authEnabled?: boolean;
    user?: string;
    password?: string;
  };
  servers: ServerConfig[];
}

export interface Schedule {
  at?: string;
}

export type HandlerCallback = (req: Express.Request, res: TypedResponse<string>, appState: AppState) => void;
