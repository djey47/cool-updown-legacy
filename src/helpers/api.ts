/**
 * Provides utility functions to handle Cool-updown API requests and responses
 */

import { ApiRequest } from "../model/models";

/**
 * @param request API request from Express middleware
 * @returns server identifier which is passed through a path parameter
 */
export function extractServerIdentifier(request: ApiRequest) {
  return request.params.serverId;
}
