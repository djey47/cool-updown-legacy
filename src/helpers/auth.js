/* @flow */

const basicAuth = require('express-basic-auth');

/*
 * Generated by:
 * https://www.random.org/strings/?num=10&len=10&digits=on&upperalpha=on&loweralpha=on&unique=on&format=html&rnd=new
 */
const AUTH_REALM = 'NyDGMFeptk';

/**
 * Basic HTTP authentication
 */
function initBasicAuth(app, isEnabled, userName, password) {
  if (!isEnabled) return;

  app.use(basicAuth({
    users: { [userName]: password },
    challenge: true,
    realm: AUTH_REALM,
  }));
}

module.exports = {
  initBasicAuth,
};
