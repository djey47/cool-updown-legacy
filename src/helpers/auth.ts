import basicAuth from 'express-basic-auth';
import config from 'config';
import { readFile } from 'fs/promises';

interface App {
  use: (plugin: unknown) => unknown;
}

/*
 * Generated by:
 * https://www.random.org/strings/?num=10&len=10&digits=on&upperalpha=on&loweralpha=on&unique=on&format=html&rnd=new
 */
const AUTH_REALM = 'NyDGMFeptk';

/**
 * Basic HTTP authentication
 */
export function initBasicAuth(app: App, isEnabled: boolean, userName: string, password: string) {
  if (!isEnabled) return;

  app.use(basicAuth({
    users: { [userName]: password },
    challenge: true,
    realm: AUTH_REALM,
  }));
}

/**
 * @private
 * @return Promise which resolves to String,
 * with the contents of SSH private key as per provided keyPath in configuration
 */
export async function readPrivateKey() {
  const privateKeyPath = config.get('server.keyPath') as string;
  const privateKey = await readFile(privateKeyPath, 'utf-8');

  // console.log('services::getPrivateKey', { privateKeyPath, privateKey });

  return privateKey as string;
}
