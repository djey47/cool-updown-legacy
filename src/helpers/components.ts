/*
 * Misc. page components
 */

/**
 * @returns contents in a pragraph then a link in a next one
 */
export function withBackLink(contents: string, linkUrl: string, linkText: string) {
    return `
    <p class="backlink-contents">${contents}</p>
    <p class="backlink-link"><a href="${linkUrl}">${linkText}</a></p>
    `;
}
