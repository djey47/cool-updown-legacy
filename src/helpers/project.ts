import appRootDir from 'app-root-dir';
import path from 'path';
import { readFile } from 'fs/promises';

export interface PackageConfig {
  name: string;
  version: string;
}

/**
 * @returns config as object, defined in the root package.json file
 */
export async function readPackageConfiguration() {
  const configFile = path.resolve(appRootDir.get(), 'package.json');
  const rawContents = await readFile(configFile, 'utf-8');
  return JSON.parse(rawContents) as PackageConfig;
}
