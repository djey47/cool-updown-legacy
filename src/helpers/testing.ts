/**
 * Unit testing support functions
 */

/**
 * @param expressResponseMock 
 * @returns an Express default response (with mocks) for unit tests
 */
export const generateDefaultResponse = (expressResponseMock) => ({
  status: c => expressResponseMock.statusMock(c),
  send: msg => expressResponseMock.sendMock(msg),
  json: ()  => expressResponseMock.jsonMock(),
});    
