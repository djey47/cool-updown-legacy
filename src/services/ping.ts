import axios from 'axios';
import config from 'config';
import lodash from 'lodash';
import { NodeSSH } from 'node-ssh';
import dateFns from 'date-fns';
import https from 'https';
import messages from '../resources/messages';
import { ping as gatewayPing } from '../helpers/systemGateway';
import logger from '../helpers/logger';
import { interpolate, toHumanDuration } from '../helpers/format';
import { getTimeDetails } from '../helpers/date';
import { AppConfig, AppState, ServerConfig } from '../model/models';
import { TypedResponse } from '../model/express';
import { readPackageConfiguration } from '../helpers/project';
import { generatePage } from '../helpers/page';
import { MetaOptions } from '../model/page';
import { getSSHParameters } from '../helpers/ssh';

const ssh = new NodeSSH();
const { cloneDeep: loCloneDeep } = lodash;
const { differenceInMilliseconds } = dateFns;

const DEFAULT_STATUS_REFRESH_SECONDS = 60;

interface Diagnostics {
  httpStatus: string;
  isHTTPSuccess: boolean;
  isPingSuccess: boolean;
  isSSHSuccess: boolean;
  pingStatus: string;
  serverUpDownTimeMessage: string;
  sshStatus: string;
  url: string;
  hostname?: string;
  scheduleStatus: string;
}

/**
 * @private
 * @return duration in human readable format
 */
function computeDuration(from: Date, to: Date) {
  const diff = differenceInMilliseconds(to, from);
  return toHumanDuration(getTimeDetails(diff));
}

/**
 * @private
 */
async function serverSSHTest(serverId: number, serverConfiguration: ServerConfig) {
  try {
    const sshClientConfig = await getSSHParameters(serverConfiguration);
    await ssh.connect(sshClientConfig);
    logger.info(`(ping:${serverId}) SSH connection OK`);
    return true;
  } catch (err) {
    logger.error(`(ping:${serverId}) SSH connection KO: ${err}`);
    return false;
  } finally {
    ssh.dispose();
  }
}

/**
 * @private
 */
async function serverHTTPTest(serverId: number, url: string) {
  try {
    const httpsAgent = new https.Agent({
      rejectUnauthorized: false,
    });
    await axios.get(url, { httpsAgent });
    logger.info(`(ping:${serverId}) HTTP connection OK`);
    return Promise.resolve(true);
  } catch (err: unknown) {
    logger.error(`(ping:${serverId}) HTTP connection KO: ${err}`);
    return Promise.resolve(false);
  }
}

function getDisplayedConfiguration(config: AppConfig) {
  const OBFUSCATED_PWD = '********';
  const displayedConfig = loCloneDeep(config) as AppConfig;
  // App password setting is obfuscated
  if (displayedConfig.app?.password) {
    displayedConfig.app.password = OBFUSCATED_PWD;
  }
  // Server password setting is obfuscated
  displayedConfig.servers.forEach((s) => {
    if (s.ssh?.password) {
      s.ssh.password = OBFUSCATED_PWD;
    }
  })

  return displayedConfig;
}

async function diagnose(config: AppConfig, appState: AppState): Promise<Diagnostics[]> {
  const {
    ping: pingMessages, status: statusMessages,
  } = messages;

  const diagPromises = config.servers.map(async (s, serverId) => {
    const {
      startedAt: serverStartedAt, stoppedAt: serverStoppedAt, isScheduleEnabled: serverScheduleEnabled
    } = appState.servers[serverId];
    const host = s.network?.hostname || undefined;
    const url = s.url || undefined;
    const hostname = s.network?.hostname || undefined;
    const isScheduleEnabled = serverScheduleEnabled || false;

    const [isPingSuccess, isSSHSuccess, isHTTPSuccess] = await Promise.all([
      gatewayPing(host),
      serverSSHTest(serverId, s),
      url ? serverHTTPTest(serverId, url) : Promise.resolve(false),
    ]);

    // Uptime/Downtime server calculation
    const now = new Date(Date.now());
    const time = computeDuration(serverStartedAt || serverStoppedAt || now, now);
    let serverUpDownTimeMessage = pingMessages.serverUndefinedTime;
    if (serverStartedAt) {
      serverUpDownTimeMessage = interpolate(pingMessages.serverUptime, { time });
    } else if (serverStoppedAt) {
      serverUpDownTimeMessage = interpolate(pingMessages.serverDowntime, { time });
    }

    return {
      isPingSuccess,
      isSSHSuccess,
      isHTTPSuccess,
      pingStatus: isPingSuccess ? statusMessages.okay : statusMessages.kayo,
      sshStatus: isSSHSuccess ? statusMessages.okay : statusMessages.kayo,
      httpStatus: isHTTPSuccess ? statusMessages.okay : statusMessages.kayo,
      serverUpDownTimeMessage,
      url,
      hostname,
      scheduleStatus: isScheduleEnabled ? statusMessages.enabled : statusMessages.disabled,
    };
  });

  return Promise.all(diagPromises);
}

const configKeySorter = (key, value) =>
  value instanceof Object && !(value instanceof Array) ?
    Object.keys(value)
      .sort()
      .reduce((sorted, key) => {
        sorted[key] = value[key];
        return sorted
      }, {}) :
    value;

/**
 * Returns diagnostics with status message
 */
export default async function ping(_req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  const {
    startedAt,
  } = appState;
  const {
    ping: pingMessages,
  } = messages;

  // Prepare configuration to be displayed
  const appConfig = config as unknown as AppConfig;
  const displayedConfig = getDisplayedConfiguration(appConfig);

  // Diagnostics
  const diags = await diagnose(appConfig, appState);
  const serverDiagnostics = formatDiagnostics(diags);

  // Uptime app calculation
  const now = new Date(Date.now());
  const uptime = computeDuration(startedAt || now, now);

  // Refresh interval
  const refreshInterval = appConfig.ui?.statusRefreshInterval == undefined ? DEFAULT_STATUS_REFRESH_SECONDS : appConfig.ui?.statusRefreshInterval;

  // Package configuration
  const packageConfig = await readPackageConfiguration();

  const htmlResult = `
  <h1>${pingMessages.statusTitle}</h1>
  <p><em>${interpolate(pingMessages.appUptime, { uptime, refreshInterval })}</em></p>
  <ul>
  ${serverDiagnostics}
  </ul>
  <p>${pingMessages.instructions}</p>
  <h2>${pingMessages.configurationTitle}</h2>
  <pre>${JSON.stringify(displayedConfig, configKeySorter, 2)}</pre>
  <hr/>
  <section><p><em>${packageConfig.name}, v${packageConfig.version}</em></p></section>
  `;

  const metaOptions = buildMetaOptions(appConfig);

  res.send(generatePage(htmlResult, metaOptions));
}

function resolveBaseServiceUrl(serverId: number) {
  return `/${serverId}/`;
}

function resolvePowerServiceUrls(serverId: number) {
  const powerUrlBase = resolveBaseServiceUrl(serverId);
  return {
    onUrl: `${powerUrlBase}on`,
    offUrl: `${powerUrlBase}off`,
  };
}

function resolveScheduleServiceUrls(serverId: number) {
  const powerUrlBase = resolveBaseServiceUrl(serverId);
  return {
    enableScheduleUrl: `${powerUrlBase}enable`,
    disableScheduleUrl: `${powerUrlBase}disable`,
  };
}

function buildMetaOptions(appConfig: AppConfig): MetaOptions {
  const statusRefreshInterval = appConfig.ui?.statusRefreshInterval;
  return {
    refreshSeconds: statusRefreshInterval === undefined ? DEFAULT_STATUS_REFRESH_SECONDS : statusRefreshInterval,
  };
}

function formatDiagnostics(diags: Diagnostics[]) {
  const {
    ping: pingMessages,
  } = messages;

  return diags.map((d, serverId) => {
    const { httpStatus, pingStatus, sshStatus, isPingSuccess, isSSHSuccess, serverUpDownTimeMessage, url, hostname, scheduleStatus } = d;

    const powerUrls = resolvePowerServiceUrls(serverId);
    const scheduleUrls = resolveScheduleServiceUrls(serverId);

    return `
      <li>
        <h2>${interpolate(pingMessages.serverTitle, { serverId })}${hostname && (': ' + hostname) || ''}</h2>
        <p>${!isPingSuccess && !isSSHSuccess ? pingMessages.offline : ''}</p>
        <ul>
          <li>${interpolate(pingMessages.scheduleItem, { scheduleStatus })}</li>
          <li>${serverUpDownTimeMessage}</li>
          <li>${interpolate(pingMessages.pingItem, { pingStatus })}</li>
          <li>${interpolate(pingMessages.sshItem, { sshStatus })}</li>
          <li>${interpolate(pingMessages.httpItem, { httpStatus, url })}</li>
          <li>${interpolate(pingMessages.actionsItem, { ...powerUrls, ...scheduleUrls })}</li>
        </ul>
      </li>
    `;
  }).join('\n');
}
