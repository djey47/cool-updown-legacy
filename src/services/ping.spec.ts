import globalMocks from '../../config/jest/globalMocks';
import resetMocks from '../../config/jest/resetMocks';
import { generateDefaultAppState, generateDefaultResponse } from '../helpers/testing';
import ping from './ping';
import { readPackageConfiguration } from '../helpers/project';
import { FeatureStatus } from '../model/models';

jest.mock('../helpers/systemGateway', () => globalMocks.systemGatewayMock);
jest.mock('../helpers/project', () => ({
  readPackageConfiguration: jest.fn(),
}));
jest.mock('../helpers/page', () => globalMocks.pageMock);
jest.mock('../helpers/ssh', () => globalMocks.sshMock);

const {
  axiosMock, expressResponseMock, systemGatewayMock: mockSystemGateway, nodesshMock, pageMock, sshMock
} = globalMocks;

const readPackageConfigurationMock = readPackageConfiguration as jest.Mock;

const appState = generateDefaultAppState();
const res = generateDefaultResponse(expressResponseMock);

describe('ping service', () => {
  beforeEach(() => {
    resetMocks();
    readPackageConfigurationMock.mockReset();

    // Ping OK
    mockSystemGateway.pingMock.mockResolvedValue(true);

    // HTTP OK
    axiosMock.get.mockImplementation(() => Promise.resolve());

    // Package config utility
    readPackageConfigurationMock.mockResolvedValue({ name: 'cud', version: 'test' });

    // Page helper
    pageMock.generatePage.mockImplementation((html, metaOptions) => `<page-shared with meta ${JSON.stringify(metaOptions)} />${html}`);

    // SSH helper
    sshMock.getSSHParameters.mockResolvedValue({
      host: 'Host Name',
      port: 22,
      privateKey: '=== PRIVATE KEY ===',
      username: 'User',
    });

  });

  it('should perform diagnosis and send appropriate response when schedule enabled', async () => {
    // given
    const state = {
      ...appState,
    };

    // when
    await ping({}, expressResponseMock, state);

    // then
    expect(mockSystemGateway.pingMock).toHaveBeenCalled();
    expect(mockSystemGateway.pingMock.mock.calls[0][0]).toEqual('Host Name');
    expect(nodesshMock.connect).toHaveBeenCalledWith({
      host: 'Host Name',
      port: 22,
      privateKey: '=== PRIVATE KEY ===',
      username: 'User',
    });
    expect(nodesshMock.dispose).toHaveBeenCalled();
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
    expect(state.servers[0].lastPingStatus).toBe(FeatureStatus.OK); 
  });

  it('should send appropriate response when serverStartedAt defined', async () => {
    // given
    const state = {
      ...appState,
      serverStartedAt: new Date(Date.now()),
    };

    // when
    await ping({}, expressResponseMock, state);

    // then
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should send appropriate response when serverStoppedAt defined', async () => {
    // given
    const state = {
      ...appState,
      serverStoppedAt: new Date(Date.now()),
    };

    // when
    await ping({}, expressResponseMock, state);

    // then
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should send appropriate response when server ping KO', async () => {
    // given
    const state = {
      ...appState,
    };
    // Ping KO
    mockSystemGateway.pingMock.mockResolvedValue(false);

    // when
    await ping({}, expressResponseMock, state);

    // then
    expect(mockSystemGateway.pingMock).toHaveBeenCalled();
    expect(nodesshMock.connect).toHaveBeenCalled();
    expect(nodesshMock.dispose).toHaveBeenCalled();
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
    expect(state.servers[0].lastPingStatus).toBe(FeatureStatus.KO);
  });

  it('should send appropriate response when server SSH KO', async () => {
    // given
    const state = {
      ...appState,
    };
    // SSH KO
    nodesshMock.connect.mockImplementation(() => Promise.reject());

    // when
    await ping({}, expressResponseMock, state);

    // then
    expect(mockSystemGateway.pingMock).toHaveBeenCalled();
    expect(nodesshMock.connect).toHaveBeenCalled();
    expect(nodesshMock.dispose).toHaveBeenCalled();
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should send appropriate response when server HTTP KO', async () => {
    // given
    const state = {
      ...appState,
    };
    // HTTP KO
    axiosMock.get.mockImplementation(() => Promise.reject());

    // when
    await ping({}, res, state);

    // then
    expect(mockSystemGateway.pingMock).toHaveBeenCalled();
    expect(nodesshMock.connect).toHaveBeenCalled();
    expect(axiosMock.get).toHaveBeenCalledWith('http://localhost');
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should send appropriate response when server OFFLINE', async () => {
    // given
    const state = {
      ...appState,
    };
    // Ping KO
    mockSystemGateway.pingMock.mockImplementation(() => Promise.resolve(false));
    // SSH KO
    nodesshMock.connect.mockImplementation(() => Promise.reject());
    // HTTP KO
    axiosMock.get.mockImplementation(() => Promise.reject());

    // when-then
    await ping({}, res, state);

    // then
    expect(expressResponseMock.sendMock).toHaveBeenCalled();
    expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
  });
});
