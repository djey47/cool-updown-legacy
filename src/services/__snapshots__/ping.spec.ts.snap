// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ping service should perform diagnosis and send appropriate response when schedule enabled 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p></p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>OK</b></li>
          <li>SSH test: <b>OK</b></li>
          <li>HTTP test: <b>OK</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;

exports[`ping service should send appropriate response when server HTTP KO 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p></p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>OK</b></li>
          <li>SSH test: <b>OK</b></li>
          <li>HTTP test: <b>KO</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;

exports[`ping service should send appropriate response when server OFFLINE 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p>Server is likely to be OFFLINE!</p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>KO</b></li>
          <li>SSH test: <b>KO</b></li>
          <li>HTTP test: <b>KO</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;

exports[`ping service should send appropriate response when server SSH KO 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p></p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>OK</b></li>
          <li>SSH test: <b>KO</b></li>
          <li>HTTP test: <b>OK</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;

exports[`ping service should send appropriate response when server ping KO 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p></p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>KO</b></li>
          <li>SSH test: <b>OK</b></li>
          <li>HTTP test: <b>OK</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;

exports[`ping service should send appropriate response when serverStartedAt defined 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p></p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>OK</b></li>
          <li>SSH test: <b>OK</b></li>
          <li>HTTP test: <b>OK</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;

exports[`ping service should send appropriate response when serverStoppedAt defined 1`] = `
"<page-shared />
  <h1>coolupdown alive and running!</h1>
  <p><em>Running for <b>1 hour(s)</b>.</em></p>
  <ul>
  
      <li>
        <h2>Server #0: Host Name</h2>
        <p></p>
        <ul>
          <li>Schedule: <b>DISABLED</b></li>
          <li>Last start/stop attempt: <b>N/A</b></li>
          <li>Ping test: <b>OK</b></li>
          <li>SSH test: <b>OK</b></li>
          <li>HTTP test: <b>OK</b> <a href="http://localhost" target="_blank">URL</a></li>
        </ul>
      </li>
    
  </ul>
  <p>See <a href="/logs" target="_blank">logs</a> for details.</p>
  <h2>Configuration</h2>
  <pre>{
  "app": {
    "port": 4600,
    "authEnabled": true,
    "user": "bob",
    "password": "********"
  },
  "servers": [
    {
      "url": "http://localhost",
      "network": {
        "macAddress": "FF:FF:FF:FF:FF:FF",
        "broadcastIpAddress": "255.255.255.255",
        "hostname": "Host Name"
      },
      "ssh": {
        "user": "User",
        "password": "********",
        "offCommand": "OFF Command",
        "port": 22,
        "keyPath": "Key Path"
      },
      "schedule": {
        "enabled": false,
        "on": {
          "at": "00:00"
        },
        "off": {
          "at": "01:00"
        }
      }
    }
  ],
  "ui": {
    "theme": "dark"
  }
}</pre>
  <hr/>
  <section><p><em>cud, vtest</em></p></section>
  "
`;
